### PR TITLE
Do not try to serialize raw aggregations dict.

### LINKED ISSUE #11772: TypeError: Object of type FrozenEvent is not JSON serializable
When testing out commit d1e6333f12d8a121c649c6176e0ac5e915345366 on matrix.org today, we ended up with the following stacktrace:

<details>

<summary>stacktrace</summary>

```
2022-01-19 13:31:19,225 - twisted - 279 - CRITICAL - sentinel - Unhandled error in Deferred:
2022-01-19 13:31:19,237 - twisted - 279 - CRITICAL - sentinel - 
Capture point (most recent call last):
  File "/usr/local/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/synapse/src/synapse/app/generic_worker.py", line 514, in <module>
    main()
  File "/home/synapse/src/synapse/app/generic_worker.py", line 510, in main
    start(sys.argv[1:])
  File "/home/synapse/src/synapse/app/generic_worker.py", line 505, in start
    _base.start_worker_reactor("synapse-generic-worker", config)
  File "/home/synapse/src/synapse/app/_base.py", line 126, in start_worker_reactor
    run_command=run_command,
  File "/home/synapse/src/synapse/app/_base.py", line 179, in start_reactor
  File "/home/synapse/src/synapse/app/_base.py", line 163, in run
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/base.py", line 1318, in run
    self.mainLoop()
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/base.py", line 1328, in mainLoop
    reactorBaseSelf.runUntilCurrent()
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/base.py", line 967, in runUntilCurrent
    f(*a, **kw)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/defer.py", line 701, in errback
    self._startRunCallbacks(fail)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/defer.py", line 764, in _startRunCallbacks
    self._runCallbacks()
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/defer.py", line 859, in _runCallbacks
    current.result, *args, **kwargs
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/defer.py", line 1751, in gotResult
    current_context.run(_inlineCallbacks, r, gen, status)
Traceback (most recent call last):
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/internet/defer.py", line 1658, in _inlineCallbacks
    cast(Failure, result).throwExceptionIntoGenerator, gen
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/python/failure.py", line 500, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/synapse/src/synapse/http/server.py", line 779, in _async_write_json_to_request_in_thread
    json_str = await defer_to_thread(request.reactor, encode, span)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/python/threadpool.py", line 238, in inContext
    result = inContext.theWork()  # type: ignore[attr-defined]
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/python/threadpool.py", line 255, in <lambda>
    ctx, func, *args, **kw
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/synapse/env-py37/lib/python3.7/site-packages/twisted/python/context.py", line 83, in callWithContext
    return func(*args, **kw)
  File "/home/synapse/src/synapse/logging/context.py", line 958, in g
    return f(*args, **kwargs)
  File "/home/synapse/src/synapse/http/server.py", line 772, in encode
    res = json_encoder(json_object)
  File "/home/synapse/src/synapse/http/server.py", line 663, in _encode_json_bytes
    return json_encoder.encode(json_object).encode("utf-8")
  File "/usr/local/lib/python3.7/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/home/synapse/src/synapse/util/__init__.py", line 67, in _handle_frozendict
    "Object of type %s is not JSON serializable" % obj.__class__.__name__
TypeError: Object of type FrozenEvent is not JSON serializable
```

</details>

Sentry link for those with access: https://sentry.matrix.org/sentry/synapse-matrixorg/issues/239798/	

This issue only occurs on our `client_reader` workers, but does occur frequently across multiple instances of it.

Unfortunately no surrounding processed request lines are relevant. However, there do seem to be a number of replication related lines, so that may be a clue...

Note that a similar error has occurred before: https://github.com/matrix-org/synapse/issues/8678.

### PR DESCRIPTION
**Pull request by clokep**
Fixes #11772 -- in the `/context` API we were pushing the raw aggregations into the result dictionary instead of just using it to serialize the events. This is due to some questionable re-use of dictionaries that I plan to clean-up separately.

Also expands tests to cover edits under more APIs.

### CODE REVIEW

**Comment by clokep** [comment]
From the linked Sentry I also realized that we seem to be returning a (blank) entry for every event from `get_bundled_aggregations` instead of only returning events where there is something to bundle. This isn't "wrong" per-say, but it needlessly generates some dictionaries and such. I wanted to keep this PR small though so didn't change that logic here. I'll do that in a follow-up.

File: tests/rest/client/test_relations.py [hunk 1]
```diff
@@ -486,12 +487,13 @@ def test_bundled_aggregations(self):
         self.assertEquals(200, channel.code, channel.json_body)
         thread_2 = channel.json_body["event_id"]
 
-        def assert_bundle(actual):
+        def assert_bundle(event_json: JsonDict) -> None:
             """Assert the expected values of the bundled aggregations."""
+            relations_dict = event_json["unsigned"].get("m.relations")
 
             # Ensure the fields are as expected.
             self.assertCountEqual(
```
Comment (reivilibre) [inline tests/rest/client/test_relations.py:498]: this assertion function always gets me with the misleading name :face_with_spiral_eyes:

**Review by reivilibre — APPROVED** [review]
Seems good to me

File: tests/rest/client/test_relations.py (hunk 1 above)
Comment (clokep) [reply tests/rest/client/test_relations.py:498]: Definitely a confusing name. 🤷

**Pull request merged**